android/android-activity/src/Android/HaskellActivity.hsc

```diff
@@ -22,6 +25,35 @@ newtype HaskellActivity = HaskellActivity { unHaskellActivity :: Ptr HaskellActi
 
 foreign import ccall unsafe "HaskellActivity_get" getHaskellActivity :: IO HaskellActivity
 
+foreign import ccall unsafe "HaskellActivity_getFilesDir" getFilesDirCString
+  :: HaskellActivity
+  -> IO CString
+
+foreign import ccall unsafe "HaskellActivity_getCacheDir" getCacheDirCString
+  :: HaskellActivity
+  -> IO CString
+
+-- | Copy a C string into Haskell returning 'Nothing' if it is NULL.
+peekMaybeCString :: CString -> IO (Maybe String)
+peekMaybeCString str =
+  if str == nullPtr
+  then return Nothing
+  else Just <$> peekCString str
+
+-- | Get the "internal" storage directory for the app. This is where data local
+-- to the app should be stored. Note that 'Nothing' is returned if the activity
+-- is not fully initalized. In practice this means you probably need to call
+-- this inside the main widget.
+getFilesDir :: HaskellActivity -> IO (Maybe String)
```

Comment:
`Maybe FilePath`?

Reply:
Sad that `FilePath` isn't even a `newtype` but done.